File: backend-py/app/routers/activities.py
"""Activity feed + provenance endpoints.

  GET /api/activities?type=&cursor=&limit=   global, keyset-paginated feed
  GET /api/nft/{mint}/history                 full chain of custody for one NFT

The feed is the natural home for HypeChain's differentiator: a verified physical
item AND its full on-chain ownership chain, which neither StockX/GOAT (no
provenance) nor a pure-NFT marketplace (no physical authentication) can show
together. ``/nft/{mint}/history`` exposes that chain directly.

An empty result is a normal 200 (``activities: []``) — at launch, before there's
real volume, the frontend renders an honest "no activity yet" state rather than
seeded/fake data.
"""

from __future__ import annotations

from fastapi import APIRouter, HTTPException, Query

from app.schemas.activity import (
    ActivityFeedResponse,
    NftHistoryResponse,
    to_activity_item,
)
from app.services import activity as activity_service

router = APIRouter()

_VALID_TYPES = activity_service.VALID_EVENT_TYPES


@router.get("/activities", response_model=ActivityFeedResponse)
async def list_activities(
    type: str | None = Query(default=None, description="mint | listing | sale | transfer"),
    cursor: str | None = Query(default=None, description="opaque keyset cursor from nextCursor"),
    limit: int = Query(default=activity_service.DEFAULT_FEED_LIMIT, ge=1, le=activity_service.MAX_LIMIT),
) -> ActivityFeedResponse:
    """Global activity feed, newest first, keyset-paginated."""
    if type is not None and type not in _VALID_TYPES:
        raise HTTPException(
            status_code=400,
            detail=f"invalid type '{type}' — expected one of {sorted(_VALID_TYPES)}",
        )
    try:
        rows, next_cursor = await activity_service.feed(event_type=type, cursor=cursor, limit=limit)
    except ValueError as exc:
        # Malformed cursor — a client paging error, not a server fault.
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    return ActivityFeedResponse(
        activities=[to_activity_item(r) for r in rows],
        nextCursor=next_cursor,
        hasMore=next_cursor is not None,
    )


@router.get("/nft/{mint}/history", response_model=NftHistoryResponse)
async def nft_history(mint: str) -> NftHistoryResponse:
    """Full chain of custody for one NFT (provenance). Empty list if unknown mint."""
    rows = await activity_service.history(mint)
    return NftHistoryResponse(
        nftMintAddress=mint,
        activities=[to_activity_item(r) for r in rows],
    )

File: backend/src/routes/activities.js
/**
 * Activity feed + provenance endpoints (ported from FastAPI).
 *
 *   GET /api/activities?type=&cursor=&limit=   global, keyset-paginated feed
 *   GET /api/nft/:mint/history                  full chain of custody for one NFT
 *
 * The feed surfaces HypeChain's differentiator: a verified physical item AND its
 * full on-chain ownership chain. An empty result is a normal 200
 * (`activities: []`) so the frontend renders an honest "no activity yet" state
 * rather than seeded data.
 */

import express from 'express';

import * as activityService from '../services/activity.js';

const router = express.Router();

router.get('/activities', async (req, res) => {
  const { type, cursor } = req.query;

  if (type != null && !activityService.VALID_EVENT_TYPES.has(type)) {
    return res.status(400).json({
      success: false,
      error: `invalid type '${type}' — expected one of ${[...activityService.VALID_EVENT_TYPES].sort().join(', ')}`,
    });
  }

  const limit = req.query.limit != null ? Number(req.query.limit) : activityService.DEFAULT_FEED_LIMIT;
  if (req.query.limit != null && (!Number.isFinite(limit) || limit < 1)) {
    return res.status(400).json({ success: false, error: 'limit must be a positive integer' });
  }

  try {
    const { rows, nextCursor } = await activityService.feed({
      eventType: type ?? null,
      cursor: cursor ?? null,
      limit,
    });
    return res.json({
      activities: rows.map(activityService.toActivityItem),
      nextCursor,
      hasMore: nextCursor != null,
    });
  } catch (err) {
    if (err.message === 'malformed cursor') {
      // A client paging error, not a server fault.
      return res.status(400).json({ success: false, error: 'malformed cursor' });
    }
    console.error('[activities] feed error:', err);
    return res.status(500).json({ success: false, error: 'Failed to load activity feed' });
  }
});

router.get('/nft/:mint/history', async (req, res) => {
  const { mint } = req.params;
  try {
    const rows = await activityService.history(mint);
    return res.json({
      nftMintAddress: mint,
      activities: rows.map(activityService.toActivityItem),
    });
  } catch (err) {
    console.error('[activities] nft-history error:', err);
    return res.status(500).json({ success: false, error: 'Failed to load NFT history' });
  }
});

export default router;
